### PR TITLE
Add EventsMode modality

### PR DESCRIPTION
**Pull request by darthsiroftardis**
This PR introduces the `EventsMode` modality and allows for users to select between three options

1. NoEvents: This is the default operation
2. CEP47: The legacy port of CEP-47 to the CEP-78 contract
3. CES: A new Casper Event Standard created proposed by @mssteuer and the MAKE team

The events mode is supplied as an optional u8 runtime argument during installation. Extent contracts upgrading from 1.1 will be able to also supply the events mode argument to enable event tracking from that Contract version

### CODE REVIEW

File: contract/src/constants.rs [hunk 1]
```diff
@@ -70,6 +72,10 @@ pub const PAGE_LIMIT: &str = "page_limit";
 pub const UNMATCHED_HASH_COUNT: &str = "unmatched_hash_count";
 pub const MIGRATION_FLAG: &str = "migration_flag";
 pub const REPORTING_MODE: &str = "reporting_mode";
+pub const EVENTS: &str = "events";
+pub const EVENT_ID_TRACKER: &str = "id_tracker";
```
Comment (gRoussac) [inline contract/src/constants.rs]:
```suggestion
```

Does not seem to be used anywhere except `init`

File: contract/src/main.rs [hunk 2]
```diff
@@ -20,9 +20,15 @@ use alloc::{
     vec,
     vec::Vec,
 };
-use core::convert::TryInto;
+
+use constants::{ARG_EVENTS_MODE, EVENTS, EVENTS_MODE, EVENT_ID_TRACKER};
```
Comment (gRoussac) [inline contract/src/main.rs]:
```suggestion
use constants::{ARG_EVENTS_MODE, EVENTS, EVENTS_MODE};
```
as per previous comment

File: contract/src/main.rs [hunk 3]
```diff
@@ -347,14 +364,18 @@ pub extern "C" fn init() {
         .unwrap_or_revert_with(NFTCoreError::FailedToCreateDictionary);
     storage::new_dictionary(PAGE_TABLE)
         .unwrap_or_revert_with(NFTCoreError::FailedToCreateDictionary);
+    storage::new_dictionary(EVENTS).unwrap_or_revert_with(NFTCoreError::FailedToCreateDictionary);
+    storage::new_dictionary(EVENT_ID_TRACKER)
+        .unwrap_or_revert_with(NFTCoreError::FailedToCreateDictionary);
```
Comment (gRoussac) [inline contract/src/main.rs]:
```suggestion
```

as per previous comment, does not seem to be used else where

File: contract/src/error.rs [hunk 4]
```diff
@@ -142,6 +142,9 @@ pub enum NFTCoreError {
     InvalidAccessKeyName = 137,
     InvalidCheckForUpgrade = 138,
     InvalidNamedKeyConvention = 139,
+    InvalidTokenEvent = 140,
```
Comment (gRoussac) [inline contract/src/error.rs]:
```suggestion
```

Does not seem to be used except Readme

File: contract/src/error.rs [hunk 5]
```diff
@@ -142,6 +142,9 @@ pub enum NFTCoreError {
...
+    MissingEventsMode = 141,
+    InvalidEventsMode = 142,
```
Comment (gRoussac) [inline contract/src/error.rs]:
```suggestion
    MissingEventsMode = 140,
    InvalidEventsMode = 141,
```
As per previous comment, this change probably triggers some tests

File: README.md [hunk 6]
```diff
@@ -685,144 +687,147 @@ Token identifiers are stored under the `CLType` `Any` and the encoding depends o
...
+| 133  | ExceededMaxTotalSupply            |
+| 134  | MissingCep78PackageHash           |
+| 135  | InvalidCep78InvalidHash           |
+| 136  | InvalidPackageHashName            |
+| 137  | InvalidAccessKeyName              |
+| 138  | InvalidCheckForUpgrade            |
+| 139  | InvalidNamedKeyConvention         |
+| 140  | InvalidTokenEvent                 |
+| 141  | MissingEventsMode                 |
+| 142  | InvalidEventsMode                 |
```
Comment (gRoussac) [inline README.md:833]:
```suggestion
| 140  | MissingEventsMode                 |
| 141  | InvalidEventsMode                 |
```

InvalidTokenEvent does not seem to be used

File: contract/src/lib.rs [hunk 7]
```diff
@@ -5,3 +5,7 @@ extern crate alloc;
 pub mod error;
 pub mod events;
 pub mod modalities;
+pub mod constants;
+
+#[cfg(feature = "contract-support")]
```
Comment (gRoussac) [inline contract/src/lib.rs:10]: Could be useful to document this line, what does it bring ?

File: rust-toolchain [hunk 8]
```diff
@@ -1 +1 @@
-nightly-2022-06-09
+nightly-2023-01-09
```
Comment (gRoussac) [inline rust-toolchain]: @darthsiroftardis I don't pass `make lint` nor `make check-lint` with this toolchain.

```
error[E0514]: found crate `cfg_if` compiled by an incompatible version of rustc
   --> /opt/rust/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.8/src/lib.rs:187:1

```
with 

```
rustc 1.68.0-nightly (cc47b0699 2023-01-08)
```

Can you please check ?

**Review by gRoussac — CHANGES_REQUESTED** [review]
- Please check if some const could be removed as per comments
- Please check linting 

Cheers

File: contract/src/main.rs [hunk 9]
```diff
@@ -799,12 +868,31 @@ pub extern "C" fn approve() {
...
+            casper_event_standard::emit(Approval::new(token_owner_key, operator, token_identifier))
+        },
+    };
 }
 
-// This is an extremely gas intensive operation. DO NOT invoke this
-// entrypoint unless absoluteness necessary
 // Approves the specified operator for transfer token_owner's tokens.
 #[no_mangle]
 pub extern "C" fn set_approval_for_all() {
```
Comment (davidatwhiletrue) [inline contract/src/main.rs:882]: At the end of this function, a CES event is emitted always, regardless of the event mode in the contract.

**Review by davidatwhiletrue — CHANGES_REQUESTED** [review]

File: contract/src/main.rs (hunk 9 above)
Comment (gRoussac) [reply contract/src/main.rs:882]: Good catch! this means there is no negative case testing (set no events or cep47 and test no CES event has been emited and Vice versa )

Comment (darthsiroftardis) [reply contract/src/main.rs:882]: Good catch, I shall add those tests

File: contract/src/constants.rs (hunk 1 above)
Comment (darthsiroftardis) [reply contract/src/constants.rs]: Removed unneeded constant

File: contract/src/main.rs (hunk 2 above)
Comment (darthsiroftardis) [reply contract/src/main.rs]: Applied suggestion

File: contract/src/main.rs (hunk 3 above)
Comment (darthsiroftardis) [reply contract/src/main.rs]: Applied suggestion

File: contract/src/error.rs (hunk 4 above)
Comment (darthsiroftardis) [reply contract/src/error.rs]: Removed unnecessary error

File: rust-toolchain (hunk 8 above)
Comment (darthsiroftardis) [reply rust-toolchain]: Reverted to an older toolchain to allow for linting

File: contract/src/main.rs (hunk 9 above)
Comment (darthsiroftardis) [reply contract/src/main.rs:882]: Confirmed with tests, that the approval event is not emitted if NoEvents is toggled on

**Review by gRoussac — APPROVED** [review]

File: README.md (hunk 6 above)
Comment (gRoussac) [reply README.md:833]: @darthsiroftardis I believe this readme still needs update removing `InvalidTokenEvent`

**Pull request merged**